src/main/java/com/github/manolo8/darkbot/core/objects/swf/FlashMap.java

```diff
@@ -0,0 +1,379 @@
+                break;
+            }
+
+            long valueAtom = ByteUtils.getLong(BUFFER, (offset += 8));
+            AtomKind valueKind = AtomKind.of(valueAtom);
+            if (valueKind != this.valueKind) {
+                System.out.println("Invalid valueKind! expected: " + this.valueKind + ", read: " + valueKind);
+                break;
+            }
+
+            Entry entry = entries[realSize];
+            if (entry == null)
+                entry = entries[realSize] = new Entry();
+
+            entry.set(keyAtom, valueAtom);
+
+            realSize++;
+            currentSize++;
+        }
+
+        for (int i = realSize; i < this.size; i++) {
+            Entry entry = entries[i];
+            if (entry != null) entry.reset();
+        }
+
+        if (updatables != null)
+            updatables.forEach((k, wrapper) -> wrapper.resetIfNoReferences());
+
+        this.size = realSize;
+    }
+
+    @Override
+    public void clear() {
+        this.size = 0;
+        //noinspection unchecked
+        this.entries = (Entry[]) Array.newInstance(Entry.class, 0);
+        if (updatables != null)
+            updatables.forEach((k, wrapper) -> wrapper.resetReferences());
+    }
+
+    @NotNull
+    @Override
+    public Set<Map.Entry<K, V>> entrySet() {
+        return new EntrySet();
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public V get(Object key) {
+        int i = indexOf(key);
+        return i == -1 ? null : entries[i].value;
+    }
+
+    @Override
+    public V put(K key, V value) {
+        if (!valueUpdatable) throw new IllegalStateException("V Type must be Updatable!");
+        if (updatables == null) updatables = new HashMap<>();
+        updatables.put(key, new ValueWrapper(value));
+
+        return value;
+    }
+
+    @Override
+    public V remove(Object key) {
+        if (updatables == null) return null;
+        ValueWrapper v = updatables.remove(key);
+        return v == null ? null : v.value;
+    }
+
+    private int indexOf(Object key) {
+        for (int i = 0; i < size(); i++) {
+            Entry entry = entries[i];
+            if (entry.getKey() == null) continue;
+            if (entry.key.equals(key)) return i;
+        }
+
+        return -1;
+    }
+
+    private class ValueWrapper {
+        private final V value;
+
+        private int references;
+
+        private ValueWrapper(V value) {
+            this.value = value;
+        }
+
+        private void resetIfNoReferences() {
+            if (references <= 0)
+                updateIfChanged((Updatable) value, 0);
+        }
+
+        private void resetReferences() {
+            references = 0;
+        }
+
+        private @Nullable ValueWrapper removeReference() {
+            references--;
+            return null;
+        }
+
+        private @NotNull ValueWrapper addReference() {
+            references++;
+            return this;
+        }
```

Comment:
these methods should just be void. there's no reason to have a method that literally always returns null, nor for it to return this